### PR TITLE
Add support for removal of messages from chats

### PR DESCRIPTION
**Pull request by sirDonovan**
Client-side part of <a>http://www.smogon.com/forums/threads/3524852/</a>

### CODE REVIEW

File: js/client-chat.js [hunk 1]
```diff
@@ -958,6 +958,13 @@
 					});
 					break;
 
+				case 'remove':
+					if (app.user.get('isStaff')) break;
+					$('.chatmessage-' + toId(row[1])).each(function() {
+						$(this).remove();
+					});
```
Comment (Zarel) [inline js/client-chat.js]: Wow, you really don't know how to jQuery, do you?

That is literally just a slow and unreadable way to write:

    $('.chatmessage-' + toId(row[1])).remove()

Comment (Zarel) [inline js/client-chat.js]: Anyway, using a raw `$` is pretty unacceptable for something as complex as PS; we'd generally use `this.$` in place of that.

File: js/battle.js [hunk 2]
```diff
@@ -5458,6 +5458,7 @@ var Battle = (function () {
 			this.log('<div class="debug"><div class="chat"><small style="color:#999">[DEBUG] ' + Tools.escapeHTML(name) + '.</small></div></div>', preempt);
 			break;
 		case 'unlink':
+		case 'remove':
```
Comment (Zarel) [inline js/battle.js]: You're not going to implement it here, too?

File: js/client-chat.js [hunk 3]
```diff
@@ -958,6 +958,13 @@
 					});
 					break;
 
+				case 'remove':
+					if (app.user.get('isStaff')) break;
```
Comment (Zarel) [inline js/client-chat.js]: Does `isStaff` exist client-side? Probably not.

Anyway, this should be a preference. We probably shouldn't force censorship on users.

File: js/client-chat.js [hunk 4]
```diff
@@ -958,6 +958,13 @@
 					});
 					break;
 
+				case 'remove':
```
Comment (Zarel) [inline js/client-chat.js]: I'm not sure we want to use as general a word as `remove`. We could repurpose `unlink` or something.

File: js/client.js [hunk 5]
```diff
@@ -845,7 +846,8 @@
 					name: name,
 					userid: toUserid(name),
 					named: named,
-					avatar: parts[3]
+					avatar: parts[3],
+					isStaff: !!+parts[4]
```
Comment (Zarel) [inline js/client.js]: Are you actually going to add `isStaff` to userdetails? Ugh. I'd say just make it a preference.

File: js/client-chat.js (hunk 3 above)
Comment (Zarel) [inline js/client-chat.js]: Default should probably hide the messages but add a small-font notice "Some spam messages were hidden, click here to restore."

**Comment by Zarel** [comment]
This is actually pretty good other than the issues I mentioned. It was exactly how I was thinking about doing it myself, so good job.

**Comment by sirDonovan** [comment]
The comment of repurposing <code>unlink</code> gave me an idea that led to this, what do you think? https://github.com/sirDonovan/Pokemon-Showdown-Client/commit/a5fa226c85a921c6c9abb097b88dbee2d511153b

**Comment by Zarel** [comment]
For the most part, yeah, that seems fine to me.

**Comment by sirDonovan** [comment]
Added changes from https://github.com/sirDonovan/Pokemon-Showdown-Client/commit/a5fa226c85a921c6c9abb097b88dbee2d511153b.

File: js/battle.js [hunk 6]
```diff
@@ -5458,6 +5458,13 @@ var Battle = (function () {
 			this.log('<div class="debug"><div class="chat"><small style="color:#999">[DEBUG] ' + Tools.escapeHTML(name) + '.</small></div></div>', preempt);
 			break;
 		case 'unlink':
+			var user = toId(args[2]) || toId(args[1]);
+			var messages = $('.chatmessage-' + user);
```
Comment (Zarel) [inline js/battle.js]: `var $messages` makes it clearer it's a jQuery collection.

File: js/battle.js [hunk 7]
```diff
@@ -5458,6 +5458,13 @@ var Battle = (function () {
...
+			messages.find('a').contents().unwrap();
+			if (args[2]) {
+				messages.css("display", "none");
```
Comment (Zarel) [inline js/battle.js]: you mean `messages.hide()`.

**Comment by sirDonovan** [comment]
Updated.

**Pull request merged**